eth/stagedsync/stage_polygon_sync.go

```diff
@@ -165,25 +171,214 @@ func SpawnPolygonSyncStage(
+	return nil
+}
+
+func UnwindPolygonSyncStage(ctx context.Context, tx kv.RwTx, u *UnwindState, cfg PolygonSyncStageCfg) error {
+	u.UnwindPoint = max(u.UnwindPoint, cfg.blockReader.FrozenBlocks()) // protect from unwind behind files
+
+	useExternalTx := tx != nil
+	if !useExternalTx {
+		var err error
+		tx, err = cfg.db.BeginRw(ctx)
+		if err != nil {
+			return err
+		}
+
+		defer tx.Rollback()
+	}
+
+	// headers
+	unwindBlock := u.Reason.Block != nil
+	if err := rawdb.TruncateCanonicalHash(tx, u.UnwindPoint+1, unwindBlock); err != nil {
+		return err
+	}
+
+	canonicalHash, err := rawdb.ReadCanonicalHash(tx, u.UnwindPoint)
+	if err != nil {
+		return err
+	}
+
+	if err = rawdb.WriteHeadHeaderHash(tx, canonicalHash); err != nil {
+		return err
+	}
+
+	// bodies
+	if err = cfg.blockWriter.MakeBodiesNonCanonical(tx, u.UnwindPoint+1); err != nil {
+		return err
+	}
+
+	// heimdall
+	if err = UnwindHeimdall(tx, u, nil); err != nil {
+		return err
 	}
 
-	if err := tx.Commit(); err != nil {
+	if err = u.Done(tx); err != nil {
 		return err
 	}
 
+	if !useExternalTx {
+		return tx.Commit()
+	}
+
 	return nil
 }
 
-func UnwindPolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
+func UnwindHeimdall(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if err := UnwindEvents(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindSpans(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindCheckpoints(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindMilestones(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func PrunePolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
-	return nil
+func UnwindEvents(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if len(unwindTypes) > 0 && !slices.Contains(unwindTypes, "events") {
+		return nil
+	}
+
+	cursor, err := tx.RwCursor(kv.BorEventNums)
+	if err != nil {
+		return err
+	}
+
+	defer cursor.Close()
+	var blockNumBuf [8]byte
+	binary.BigEndian.PutUint64(blockNumBuf[:], u.UnwindPoint+1)
+	k, v, err := cursor.Seek(blockNumBuf[:])
+	if err != nil {
+		return err
+	}
+	if k != nil {
+		// v is the encoding of the first eventId to be removed
+		eventCursor, err := tx.RwCursor(kv.BorEvents)
+		if err != nil {
+			return err
+		}
+		defer eventCursor.Close()
+		for v, _, err = eventCursor.Seek(v); err == nil && v != nil; v, _, err = eventCursor.Next() {
+			if err = eventCursor.DeleteCurrent(); err != nil {
+				return err
+			}
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	for ; err == nil && k != nil; k, _, err = cursor.Next() {
+		if err = cursor.DeleteCurrent(); err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func UnwindSpans(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if len(unwindTypes) > 0 && !slices.Contains(unwindTypes, "spans") {
```

Comment:
Similarly, check can be pushed to the outer function.

Reply:
no it cant https://github.com/erigontech/erigon/pull/11531#discussion_r1711199537

Reply:
Can't you do :

```Go
unc UnwindHeimdall(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
        if len(unwindTypes)  == 0 {
            return nil
         }
         if len(unwindTypes) > 0 &&
	if slices.Contains(unwindTypes, "events") {
	     err := UnwindEvents(tx, u, unwindTypes); 
	     if err != nil {
		return err
	     }
	}

       ....
```

Reply:
The benefit is that you have a input validation within a single function. And actually you can even remote the `unwindTypes` parameter in `UnwindEvents` since it is not used for anything else other than input validation.

Reply:
What input validation are you referring to? I don't understand.

Also the example you posted is logically wrong: `if len(unwindTypes) == 0` should not return `nil`. If it is 0 then that means all entity types should be unwound.

Reply:
Ok, yes I see. You have to perform the unwind when ` len(unwindTypes) > 0 && slices.Contains(unwindTypes, "events")` etc.

The point I am making is that you can better handle `unwindTypes` in `UnwindHeimdall` and handle the logic there to determine what needs to be unwound and what not. That way you don't need a redundant `unwindTypes` argument in `UnwindSpans`, `UnwindEvents` , etc...

Reply:
oh wait sorry, it's the opposite , you unwind if `len(unwindTypes) == 0  || slices.Contains(unwindTypes, "spans")`. I'm not sure I understand the motivation for this choice. Why would "spans" being included in `unwindTypes` indicate the spans should not be unwound?

Reply:
I don't think there is much value in decoupling it now since there are no other usages of these functions right now and there is a lot less nesting this way. If there were other usages of UnwindSpans, UnwindEvents, etc. then yes I agree with you it is preferable to decouple. However this decoupling can be done in the future if it is ever needed (hint: read about YAGNI).

Regardless, I would like to move on with this so I've changed the code to match your suggestion.

Reply:
I think there is still value to removing this parameter. Imagine some other part of code, or some RPC call wanting to call `UnwindSpans` or `UnwindEvents` . Why should they have to bother with this unnecessary extra parameter? It just adds extra complication.

And about YAGNI defined as : _( a principle which arose from extreme programming (XP) that states a programmer should not add functionality until deemed necessary)_

I wouldn't call this as adding any functionality, we are simply removing a redundant parameter.

Reply:
The word "imagine" is to be highlighted here.
Imagine is different from reality.
When such a need comes to exist, then that is the right time to decouple this.